nondex-core/src/main/java/java/util/WeakHashMap.java

```diff
@@ -775,13 +777,33 @@ public String toString() {
          */
         private Object currentKey = null;
 
+        private Iterator<Entry<K,V>> iter = null;
+
         HashIterator() {
             index = isEmpty() ? 0 : table.length;
+            List<Entry<K,V>> order = new ArrayList<>();
+            while(originalHasNext()) {
+                order.add(originalNextEntry());
+            }
+            order = edu.illinois.nondex.shuffling.ControlNondeterminism.shuffle(order);
+            iter = order.iterator();
+            lastReturned = null;
         }
 
         public boolean hasNext() {
-            Entry<K,V>[] t = table;
+            return iter.hasNext();
+        }
 
+        public Entry<K, V> nextEntry() {
+            if (modCount != expectedModCount)
+                throw new ConcurrentModificationException();
+            lastReturned = iter.next();
+            currentKey = lastReturned.get();
+            return lastReturned;
+        }
+
+        public boolean originalHasNext() {
+            Entry<K,V>[] t = table;
```

Comment:
I actually didn't modify it. You can look at line 783 and line 806 and you will find that this is because of how the "diff" works.